### PR TITLE
feat(components): Implement Interface for Managing GTM Datalayer. 

### LINKED ISSUE #451: Implement Interface for Managing GTM Meta Data
## Description
As a Content Editor, I want the ability to manage meta data via the interface, so that I do not have to bring in a developer to manage the content. 

Wires: **https://jhdfdc.axshare.com/#g=1&p=generic_ui 

*Form Title:* 

Google Tag Manager 

**Form Description Text:**

Enter the page level metadata that will be used with Google Tag Manager.  

**Form fields:**
|*Title*|*Field*|*Placeholder text to appear in field*|
|Product SKU|Text Field| |
|Product UPC|Text Field| |
|Product Name|Text Field| |
|Product Variant Name|Text Field| |
|Page Type|Text Field| |

Notes:
- No validation 
- No need for scrolling  
- Keep in mind that this form may be combined with another form such as GTM meta data at a later time 
- Long text field should be 3 lines high
  
## Acceptance Criteria
AC1: The Content Edit UI Main Menu includes a "GTM" Icon 

AC2: Clicking on the GTM" Icon opens a form panel named "Google Tag Manager" (form.png) that lists all the available data points (in the table in the description)

AC3: The data is used and stored from the GTM content implemented in #450

### PR DESCRIPTION
**Pull request by awm086**
### Changes:
- Add an API to allow for composing user interface for GTM datalayers.

### Test Instructions:
- Enabled google tag manager. 
- Visit the home page on example/test-site.
- Note the GTM form appears.
- Note the GTM form had page type field.
- Visit a product page.
- Note the product page has example properties defined such as sku, name, variant.
- Edit and save the fields.
- Disable Edit mode.
- Inspect the head element for the appropriate datalayer script with element. 
 
### Related Issues
Fixes #451

### CODE REVIEW

File: examples/test-site/.env.site [hunk 1]
```diff
@@ -22,7 +22,7 @@ SITE_URL='https://www.example.com/'
 
 # GTM is enabled by default.
 # To disable it - uncomment the following line
-# GOOGLE_TAGMANAGER_ENABLED='0'
+GOOGLE_TAGMANAGER_ENABLED='1'
```
Comment (awm086) [inline examples/test-site/.env.site:25]: remove -- used for quick testing.

File: examples/test-site/src/components/GTM/index.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+};
+
+// Define the global dataLayer default data.
+
+// Add a page type editable field which value will be injected in the default
+// dataLayer defined above at a given path.
+const withDataLayerPageType = withDataLayerItem({
+  name: 'pagetype',
+  label: 'Page Type',
+  path: 'secondObject.page.pageType',
```
Comment (dewen) [inline examples/test-site/src/components/GTM/index.tsx:59]: There's no path attribute from gtm Options type (packages/bodiless-components/src/GTM/gtm.tsx), you might need to add this to Meta Options type.

Comment (awm086) [reply examples/test-site/src/components/GTM/index.tsx:59]: @dewen withDataLayerItem takes Options of type:
```
type Options = {
  label: string;
  path?: string;
  useFormElement?: () => CT<FieldProps<any, any>>;
  placeholder?: string;
} & BasicOptions;
```

File: examples/test-site/src/components/GTM/index.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      page: {
+        country: 'US',
+        language: 'en',
+        hostname: 'www.example.com',
+      },
+    },
+  },
+};
+
+// Define the global dataLayer default data.
```
Comment (dewen) [inline examples/test-site/src/components/GTM/index.tsx:52]: Redundant line.

File: examples/test-site/src/components/GTM/index.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+const useMenuOptions = () => [
+  {
+    name: 'gtm',
+    icon: 'local_offer',
+    label: 'GTM',
+  },
+];
+
+// Define the form Title and description.
+export const gtmFormHeader = {
```
Comment (dewen) [inline examples/test-site/src/components/GTM/index.tsx:72]: Do we need export here? Seems not referred from outside the module.

File: examples/test-site/src/templates/product_detail.jsx [hunk 5]
```diff
@@ -29,7 +28,12 @@ import {
   AccordionClean,
   asTestableAccordion,
 } from '@bodiless/organisms';
-import { withEvent, asBodilessHelmet } from '@bodiless/components';
+import {
+  withDataLayerItem,
+  withDefaultDataLayer,
+} from '@bodiless/components';
+import Helmet from 'react-helmet';
```
Comment (dewen) [inline examples/test-site/src/templates/product_detail.jsx:35]: Shall we move this react-helmet before @bodiless imports for consistency?
https://github.com/benmosher/eslint-plugin-import/blob/v2.22.0/docs/rules/order.md

File: examples/test-site/src/templates/product_detail.jsx [hunk 6]
```diff
@@ -43,6 +47,70 @@ import { FlowContainerDefault } from '../components/FlowContainer';
...
+  },
+};
+
+// Define a product UPC editable field to be added to the GTM form.
+const withDataLayerSku = withDataLayerItem({
+  name: 'sku',
+  label: 'Product SKU',
+  // The path relevant the default dataLayer which is a union of page or
+  // product level data and the global site data. In this example, the product
+  // data are added to the global data array at index 2 hence the 2.product...
```
Comment (dewen) [inline examples/test-site/src/templates/product_detail.jsx:73]: Is this "index 2" comment still valid? It's "productObject.product.0" below.

File: packages/bodiless-components/src/Meta/withMetaForm.tsx [hunk 7]
```diff
@@ -42,7 +42,7 @@ export type MetaSnippetOptions = {
 
 export const withMetaSnippet = (
   options: MetaSnippetOptions,
-) => withEditFormSnippet({
+) => ifEditable(withEditFormSnippet({
```
Comment (dewen) [inline packages/bodiless-components/src/Meta/withMetaForm.tsx:45]: Why do we need ifEditable here? The ifEditable logic current applied from withMeta/withTitle.

I think it makes sense to have client decide if ifEditable should be applied here.

File: packages/bodiless-components/src/Meta/withMetaForm.tsx [hunk 8]
```diff
@@ -99,9 +99,9 @@ const defaultMetaFormHeader = {
 const withMetaForm = (
   useMenuOptions: (props: any) => TMenuOption[],
   metaFormHeader?: HeaderProps,
-) => flowRight(
+) => ifEditable(
```
Comment (dewen) [inline packages/bodiless-components/src/Meta/withMetaForm.tsx:102]: same as for withMetaSnippet, you can apply ifEditable from inside withGlobalGTM.

**Review by dewen — DISMISSED** [review]

File: packages/bodiless-components/src/GTM/gtm.tsx [hunk 9]
```diff
@@ -97,7 +97,7 @@ const withDataLayerScript = (HelmetComponent: CT<BaseProps>) => (
   props: Props,
 ) => {
   const { isEdit } = useEditContext();
-  if (!tagManagerEnabled || isEdit) {
+  if (!tagManagerEnabled || isEdit || process.env.NODE_ENV !== 'production') {
```
Comment (dewen) [inline packages/bodiless-components/src/GTM/gtm.tsx:100]: NODE_ENV is an environment variable from Express, right? Shall we introduce this dependency into bodiless components?

File: examples/test-site/src/components/GTM/productViewd.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
+/**
```
Comment (hvanyo) [inline examples/test-site/src/components/GTM/productViewd.tsx:1]: Can we fix the name of this to productViewed.tsx

File: examples/test-site/src/components/GTM/index.tsx [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+});
+
+/*
+ * Only render the script in static/production mode.
+ */
+const renderDataLayerScript = (ct : ComponentType) => {
+  if (process.env.NODE_ENV === 'production') {
+    return withDataLayerScript(ct);
+  }
+  return () => (<></>);
```
Comment (dewen) [inline examples/test-site/src/components/GTM/index.tsx:82]: You might want to return ct here if not on static site?

Comment (awm086) [reply examples/test-site/src/components/GTM/index.tsx:82]: correct - updated

**Review by dewen — APPROVED** [review]

**Review by hvanyo — APPROVED** [review]

**Pull request merged**